Account for missing keyword in fn/struct definition

estebank opened the pull request:
Fix #38911.

estebank committed:
Fix underline in suggestions

estebank committed:
Account for missing keyword in fn/struct definition

estebank commented:
r? @nikomatsakis

nikomatsakis reviewed (changes requested):
Actually, looking more closely, I think this is overkill. I'll leave a more detailed comment with an alternative shortly.

nikomatsakis commented:
Currently, after failing to parse, the parser invokes this commonly used method:

https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6267-L6275

This is intended to look for a macro use like `path!` -- but right now it decides it has seen a macro quite early, based [only on whether the next token can start a path](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6276). Presuming it can, we [error if we saw a `pub`](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6280), [parse the path](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6285),  and then [expect a `!`](https://github.com/rust-lang/rust/blob/ce2b8a4944f1983759cb7d0e8d35a8bb889dbee6/src/libsyntax/parse/parser.rs#L6286). **This is where the error will occur in our bad examples.**

I think we should customize what happens when we "expect" this `!` but do not find it:

To start, we can look at the path. If it was multipart, like `a::b`, then I think this could only be a macro-use, perhaps one where the `!` was forgotten, and we could diagnose it that way. (Or, at least, we could do this if the next token is `(`.)

If the path is a single identifier, we can then examine the next character: if it is `(`, we could suggest a function. If it is `{`, we could suggest a struct. Or, we could just issue something generic like "expected an item like a `fn` or a macro use like `foo!`".

estebank commented:
It is reasonable to go ahead and do that change, but I also feel there's value in modifying the current PR to keep doing what it is doing and returning `Ok` with the fn or struct that could be parsed, instead of the `Err`. That way parsing can continue and present other errors down the line, instead of the current behavior of having a failure parsing a macro and then a failure parsing the opening brace or paren. What do you think?

nikomatsakis commented:
@estebank that's a good point. On the other hand, I wonder, could we get a similar effect by just skipping over entire token trees?

I feel like only being able to recover when the input is **perfect** is perhaps not ideal.

nikomatsakis commented:
though man but mismatched braces are the worst sort of parse errors =)

nikomatsakis commented:
In any case, I could go either way. But I'd like to see us able to recover even from something like this:

```rust
pub foo(x: u32) {
    x->foo(); // whoops, too much C++
}
```

petrochenkov commented:
https://github.com/rust-lang/rust/pull/41282 already implemented this recovery for trait and impl items, it can be adapted to normal items as well.
Saving/restoring parser state should be avoided if possible.

estebank committed:
Do not rewind parser and ignore following blocks

When encountering `pub ident`, attempt to identify the code that comes
afterwards, wether it is a brace block (assume it is a struct), a paren
list followed by a colon (assume struct) or a paren list followed by a
block (assume a fn). Consume those blocks to avoid any further parser
errors and return a `Placeholder` item in order to allow the parser to
continue. In the case of unenclosed blocks, the behavior is the same as
it is currently: no further errors are processed.

estebank committed:
Do not attemt to continue parsing after `pub ident`

Try to identify the following code in order to provide better
diagnostics, but return the error to bail out early during the parse.

estebank commented:
I have modified the code to not perform the rewind. The outwards behavior now is similar to what it already was (the parser will not let the compiler to continue, but it will try to consume the continuing code in order to identify what it was meant (fn or struct), regardless of code errors (just looks for enclosed `(...)` and `{...}` sections).

estebank committed:
Suggest macro call when not sure that it is fn definition

nikomatsakis commented on src/test/ui/suggestions/pub-ident-fn-2.stderr:
```diff
@@ -0,0 +1,13 @@
+   |
+help: add `fn` here to parse `foo` as a public method
+   |
+11 | pub fn foo(s: usize) { bar() }
```
that is pretty slick =)

nikomatsakis commented on src/libsyntax/parse/parser.rs:
```diff
@@ -6225,7 +6275,65 @@ impl<'a> Parser<'a> {
+        // Verify wether we have encountered a struct or method definition where the user forgot to
+        // add the `struct` or `fn` keyword after writing `pub`: `pub S {}`
+        if visibility == Visibility::Public &&
+            self.check_ident() &&
```
Isn't it possible these days to have *paths* to macros? e.g., `foo::bar!`?

cc @jseyfried @petrochenkov

estebank replied:
The way the fall back happens, if the code isn't either `pub foo (` or `pub foo {`, the current macro handling will happen. In the case of `pub foo::bar` this new branch will not be executed. As long as `parse_macro_use_or_failure` does the right thing with paths, this should be fine :)

nikomatsakis reviewed (commented):
This all looks quite cool. I do have one concern about the recovery condition I'd like to get resolved though.

estebank committed:
Emit `DocComment` in bad location error but continue parsing struct fields

estebank committed:
Revert to correct recovery behavior

estebank committed:
Consume trailing doc comments to avoid parse errors

estebank commented:
@bors try

bors commented:
:hourglass: Trying commit f103342b8f63e6e88992749b5d0a49995bad5d48 with merge 1026513cc276e201297365f43014ee01b2df9b06...

estebank committed:
Add inline `compile-fail` markers to tests

estebank commented:
@nikomatsakis this PR is ballooning in size with multiple independent changes that are kind or related but not entirely. There's a very small one liner suggestion underline fix, emitting inappropriate `DocComment` location error but allow it for the parser to continue parsing the struct,  making block parsing errors recoverable, other than the original intention of the PR which was to add a suggestion if the parser encounters `pub ident (` or `pub ident {`.

Because of these changes, in the case of [`compile-fail-fulldeps/proc-macro/derive-bad.rs`](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail-fulldeps/proc-macro/derive-bad.rs) (which uses [`compile-fail-fulldeps/proc-macro/auxiliary/derive-bad.rs`](https://github.com/rust-lang/rust/blob/master/src/test/compile-fail-fulldeps/proc-macro/auxiliary/derive-bad.rs)) the parsing errors from the `derive` are emitted and not seen by `libproc_macro` as a failure, so we get the internal parsing errors, that somebody who can't modify the proc macro can't do anything about, and we don't cause the actual expected message. For the second issue I'll try to find a way to emit it, for the first, is it ok to leak these messages?

As for the first paragraph, should I try to split this in a bunch of different dependent PRs?

estebank committed:
Fix proc_macro output with struct parse error

estebank commented:
@nikomatsakis let me know if the output changes are reasonable or if I should keep working on this PR to get to minimal changes from the status quo.

nikomatsakis commented:
@bors r+

@estebank seems good to me.

bors commented:
:pushpin: Commit cf9283e has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis cf9283ea9376525c59015a52c729e7a79f576426 -->

bors commented:
:hourglass: Testing commit cf9283ea9376525c59015a52c729e7a79f576426 with merge d1364a65c0e064ee12aea8d58aceb403b8200682...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.5565), [status-travis](https://travis-ci.org/rust-lang/rust/builds/309880422?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing d1364a65c0e064ee12aea8d58aceb403b8200682 to master...

Xiretza commented on src/libsyntax/parse/parser.rs:
```diff
@@ -6236,7 +6279,65 @@ impl<'a> Parser<'a> {
+                    self.check(&token::OpenDelim(token::Brace))
+                {
+                    ("fn", "method", false)
+                } else if self.check(&token::Colon) {
```
@estebank sorry for the enormous necropost, but do you know what the idea behind this was? It matches `pub foo(stuff):` and suggests adding `struct`, but where does that colon come from? Is it possible this was intended to be `token::Semi` instead? The behaviour isn't covered by any tests, unfortunately.
